In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime

In [2]:
#Importing Dataframe
wb_location = pd.read_csv("prof-W-B-location-edited.csv")
wb_location.head(15)

Unnamed: 0.1  Unnamed: 0        lat         lon  \
0              1      909779  39.739592 -105.161785   
1              2      909780  39.767084 -105.142443   
2              3      909781  39.784075 -105.135851   
3              4      909782  39.784501 -105.085261   
4              5      909783  39.784735 -105.063591   
5              6      909784  39.784399 -105.043754   
6              7      909785  39.783046 -105.002327   
7              8      909786  39.779305 -104.983603   
8              9      909787  39.774280 -104.861977   
9             10      909788  39.770895 -104.818076   
10            11      909789  39.770895 -104.818076   
11            12      909790  39.770895 -104.818076   
12            13      909791  39.853366 -104.670565   
13            14      909792  39.848740 -104.673091   
14            15      909793  39.848418 -104.673037   

                           timestamp                          datetime  \
0   2022-03-09 12:02:03.803000+00:00  2022-03-09 12:02:03.803000+00:00   
1   2022-03-09 12:04:04.224000+00:00  2022-03-09 12:04:04.224000+00:00   
2   2022-03-09 12:05:03.769000+00:00  2022-03-09 12:05:03.769000+00:00   
3   2022-03-09 12:07:21.803000+00:00  2022-03-09 12:07:21.803000+00:00   
4   2022-03-09 12:08:21.830000+00:00  2022-03-09 12:08:21.830000+00:00   
5   2022-03-09 12:09:21.863000+00:00  2022-03-09 12:09:21.863000+00:00   
6   2022-03-09 12:11:22.090000+00:00  2022-03-09 12:11:22.090000+00:00   
7   2022-03-09 12:12:21.943000+00:00  2022-03-09 12:12:21.943000+00:00   
8   2022-03-09 12:18:22.111000+00:00  2022-03-09 12:18:22.111000+00:00   
9   2022-03-09 12:20:22.221000+00:00  2022-03-09 12:20:22.221000+00:00   
10  2022-03-09 12:31:46.740000+00:00  2022-03-09 12:31:46.740000+00:00   
11  2022-03-09 12:32:08.209000+00:00  2022-03-09 12:32:08.209000+00:00   
12  2022-03-09 12:32:27.413000+00:00  2022-03-09 12:32:27.413000+00:00   
13  2022-03-09 12:33:11.974000+00:00  2022-03-09 12:33:11.974000+00:00   
14  2022-03-09 12:34:12.029000+00:00  2022-03-09 12:34:12.029000+00:00   

                time_delta  distance_Olin  
0   0 days 00:00:59.597000    1279.600701  
1   0 days 00:02:00.421000    1279.046263  
2   0 days 00:00:59.545000    1278.344000  
3   0 days 00:02:18.034000    1281.746995  
4   0 days 00:01:00.027000    1283.202415  
5   0 days 00:01:00.033000    1284.572405  
6   0 days 00:02:00.227000    1287.479040  
7   0 days 00:00:59.853000    1289.003769  
8   0 days 00:06:00.168000    1297.622644  
9   0 days 00:02:00.110000    1300.843202  
10  0 days 00:11:24.519000    1300.843202  
11  0 days 00:00:21.469000    1300.843202  
12  0 days 00:00:19.204000    1305.460223  
13  0 days 00:00:44.561000    1305.592273  
14  0 days 00:01:00.055000    1305.617201

Chosen Location: Taqueria Mi Pueblito (Mi Puebs)!: (46.07529, -118.31327)

In [3]:
#Now let's use the haversine formula to determine Prof WB location to Mi Puebs!
#Thank you to this stack overflow user for this function: derricw! (https://stackoverflow.com/users/1390022/derricw)
def haversine_np(lon1, lat1, lon2, lat2):
    """
    Calculate the great circle distance between two points
    on the earth (specified in decimal degrees)

    All args must be of equal length.    

    """
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = np.sin(dlat/2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2.0)**2

    c = 2 * np.arcsin(np.sqrt(a))
    km = 6367 * c
    return km

In [4]:
#Implementing the Formula into WB Locational Data
wb_location["distance_Mi_Puebs"] = haversine_np(wb_location["lon"], wb_location["lat"], -118.31327, 46.07529)


In [5]:
wb_location

Unnamed: 0.1  Unnamed: 0        lat         lon  \
0                1      909779  39.739592 -105.161785   
1                2      909780  39.767084 -105.142443   
2                3      909781  39.784075 -105.135851   
3                4      909782  39.784501 -105.085261   
4                5      909783  39.784735 -105.063591   
...            ...         ...        ...         ...   
1613          1614      911392  39.786299 -104.811715   
1614          1615      911393  39.786206 -104.811759   
1615          1616      911394  39.785963 -104.811527   
1616          1617      911395  39.786378 -104.811420   
1617          1618      911396  39.785948 -104.811264   

                             timestamp                          datetime  \
0     2022-03-09 12:02:03.803000+00:00  2022-03-09 12:02:03.803000+00:00   
1     2022-03-09 12:04:04.224000+00:00  2022-03-09 12:04:04.224000+00:00   
2     2022-03-09 12:05:03.769000+00:00  2022-03-09 12:05:03.769000+00:00   
3     2022-03-09 12:07:21.803000+00:00  2022-03-09 12:07:21.803000+00:00   
4     2022-03-09 12:08:21.830000+00:00  2022-03-09 12:08:21.830000+00:00   
...                                ...                               ...   
1613  2022-03-12 22:01:34.597000+00:00  2022-03-12 22:01:34.597000+00:00   
1614  2022-03-12 22:02:06.613000+00:00  2022-03-12 22:02:06.613000+00:00   
1615  2022-03-12 22:02:42.434000+00:00  2022-03-12 22:02:42.434000+00:00   
1616  2022-03-12 22:03:14.305000+00:00  2022-03-12 22:03:14.305000+00:00   
1617  2022-03-12 22:03:50.655000+00:00  2022-03-12 22:03:50.655000+00:00   

                  time_delta  distance_Olin  distance_Mi_Puebs  
0     0 days 00:00:59.597000    1279.600701        1278.683149  
1     0 days 00:02:00.421000    1279.046263        1278.126570  
2     0 days 00:00:59.545000    1278.344000        1277.423160  
3     0 days 00:02:18.034000    1281.746995        1280.824500  
4     0 days 00:01:00.027000    1283.202415        1282.279212  
...                      ...            ...                ...  
1613  0 days 00:00:33.481000    1300.248459        1299.317239  
1614  0 days 00:00:32.016000    1300.251709        1299.320496  
1615  0 days 00:00:35.821000    1300.283733        1299.352526  
1616  0 days 00:00:31.871000    1300.263328        1299.332096  
1617  0 days 00:00:36.350000    1300.302705        1299.371490  

[1618 rows x 9 columns]

In [6]:
#Now let's determine how close professor WB got to Mi-Puebs and the subesquent Timestamp of this moment!
wb_location[wb_location["distance_Mi_Puebs"] == wb_location["distance_Mi_Puebs"].min()]


Unnamed: 0.1  Unnamed: 0        lat         lon  \
308           309      910087  46.067667 -118.312772   

                            timestamp                          datetime  \
308  2022-03-09 23:21:58.530000+00:00  2022-03-09 23:21:58.530000+00:00   

                 time_delta  distance_Olin  distance_Mi_Puebs  
308  0 days 00:00:40.112000       1.359732           0.847976

We now know the closest Professor W-B got to Mi Puebs was 0.84 km away from the restuarant at 23:21:58.530000+00:00 on 2022-03-09. 

In [7]:
#What about how far away? 
wb_location[wb_location["distance_Mi_Puebs"] == wb_location["distance_Mi_Puebs"].max()]

Unnamed: 0.1  Unnamed: 0        lat         lon  \
1545          1546      911324  39.842509 -104.675727   

                             timestamp                          datetime  \
1545  2022-03-12 21:34:38.645000+00:00  2022-03-12 21:34:38.645000+00:00   

                  time_delta  distance_Olin  distance_Mi_Puebs  
1545  0 days 00:00:39.493000    1305.822946        1304.884575

We now know the furthest away Professor W-B was to Mi Puebs was 1304.884575 km away from the restuarant at 21:34:38.645000+00:00 on 2022-03-12.

The rest of Part Two is completed via Observable!! A link to the obserable map/applicable portion to this project is here: https://observablehq.com/@bernie-jones-ws/projectsevenparttwo


Due to uncertainty where to position observations from the mapping excercise. The same response from Observable has been pasted here as well:
    
  Looking at both Professor W-B data in a domestic setting and more locally within the Walla Walla region offers some pretty intersting insights to their travels. Domestically facing this data suggests Professor WB spent some time in Seattle in addition to Colorado and Walla Walla. There also appears to be some time spent slightly east of Seattle (roughly) suggestive of either a visit to something such as a trail or suggesting travel eastward?! 

  Shifting to local travel within Walla Walla it appears Professor WB largely spent time in two settings. Overlaying this map into a satellite version would be a further extension that would offer tremendous insight. We did this partially during an in-class exercise. It appears Professor WB spent time largely on Whitman's campus and then in downtown Walla Walla.